### PR TITLE
[office-js] [office-js-preview] (Outlook) Add missing options overloads

### PR DESCRIPTION
**Pull request by ElizabethSamuel-MSFT**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: https://docs.microsoft.com/javascript/api/outlook/office.appointmentcompose#setselecteddataasync-data--options--callback-
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/office-js-preview/index.d.ts [hunk 1]
```diff
@@ -10928,6 +10928,32 @@ declare namespace Office {
...
+         * </table>
+         * 
+         * @param data - The string to be inserted at the beginning of the body. The string is limited to 1,000,000 characters.
+         * @param options - Optional. An object literal that contains one or more of the following properties.
+         *        asyncContext: Developers can provide any object they wish to access in the callback method.
+         *        coercionType: The desired format for the body. The string in the data parameter will be converted to this format.
+         * @param callback - Optional. When the method completes, the function passed in the callback parameter is called with a single parameter of type Office.AsyncResult.
+         *                  Any errors encountered will be provided in the asyncResult.error property.
+         */
+        prependAsync(data: string, options?: CoercionTypeOptions): void;
```
Comment (amcasey) [inline types/office-js-preview/index.d.ts]: Does this make the preceding overload redundant?

Comment (AlexJerabek) [reply types/office-js-preview/index.d.ts]: Agreed. With these, the preceding function is unnecessary.

File: types/office-js-preview/index.d.ts [hunk 2]
```diff
@@ -11004,6 +11030,34 @@ declare namespace Office {
...
+         * </table>
+         * 
+         * @param data - The string that will replace the existing body. The string is limited to 1,000,000 characters.
+         * @param options - Optional. An object literal that contains one or more of the following properties.
+         *        asyncContext: Developers can provide any object they wish to access in the callback method.
+         *        coercionType: The desired format for the body. The string in the data parameter will be converted to this format.
+         * @param callback - Optional. When the method completes, the function passed in the callback parameter is called with a single parameter of type Office.AsyncResult.
+         *                  Any errors encountered will be provided in the asyncResult.error property.
+         */
+        setAsync(data: string, options?: CoercionTypeOptions): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: This makes the preceding overload unnecessary.

File: types/office-js-preview/index.d.ts [hunk 3]
```diff
@@ -11081,6 +11135,31 @@ declare namespace Office {
...
+         *   <tr><td>{@link https://docs.microsoft.com/outlook/add-ins/#extension-points | Applicable Outlook mode}</td><td>Compose</td></tr>
+         *   <tr><td>Errors</td><td>DataExceedsMaximumSize - The data parameter is longer than 1,000,000 characters.</td></tr>
+         *   <tr><td></td><td>InvalidFormatError - The options.coercionType parameter is set to Office.CoercionType.Html and the message body is in plain text.</td></tr>
+         * </table>
+         * 
+         * @param data - The string that will replace the existing body. The string is limited to 1,000,000 characters.
+         * @param options - Optional. An object literal that contains one or more of the following properties.
+         *        coercionType: The desired format for the body. The string in the data parameter will be converted to this format.
+         */
+        setSelectedDataAsync(data: string, options?: CoercionTypeOptions): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: This makes the preceding overload unnecessary.

File: types/office-js-preview/index.d.ts [hunk 4]
```diff
@@ -12315,6 +12394,30 @@ declare namespace Office {
...
+         *   <tr><td></td><td>FileTypeNotSupported - The attachment has an extension that is not allowed.</td></tr>
+         *   <tr><td></td><td>NumberOfAttachmentsExceeded - The message or appointment has too many attachments.</td></tr>
+         * </table>
+         * 
+         * @param uri - The URI that provides the location of the file to attach to the message or appointment. The maximum length is 2048 characters.
+         * @param attachmentName - The name of the attachment that is shown while the attachment is uploading. The maximum length is 255 characters.
+         * @param options - Optional. An object literal that contains one or more of the following properties.
+         *        isInline: If true, indicates that the attachment will be shown inline in the message body, and should not be displayed in the attachment list.
+         */
+        addFileAttachmentAsync(uri: string, attachmentName: string, options?: { isInline: boolean }): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: This overload covered by the preceding one and is unnecessary.

File: types/office-js-preview/index.d.ts [hunk 5]
```diff
@@ -12396,6 +12499,32 @@ declare namespace Office {
...
+         * </table>
+         * 
+         * @param base64File - The base64 encoded content of an image or file to be added to an email or event.
+         * @param attachmentName - The name of the attachment that is shown while the attachment is uploading. The maximum length is 255 characters.
+         * @param options - Optional. An object literal that contains one or more of the following properties.
+         *        isInline: If true, indicates that the attachment will be shown inline in the message body and should not be displayed in the attachment list.
+         * 
+         * @beta
+         */
+        addFileAttachmentFromBase64Async(base64File: string, attachmentName: string, options?: { isInline: boolean }): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: This and the preceding method are unnecessary.

**Review by AlexJerabek — CHANGES_REQUESTED** [review]
Left some comments, but they're all the same pattern. Basically, the overloads you need are:
- foo(reqParameter, options?: OptionType, callback?: Callback)
- foo(reqParameter, callback?: Callback)

File: types/office-js-preview/index.d.ts [hunk 6]
```diff
@@ -17134,38 +15567,10 @@ declare namespace Office {
...
-         * @param options - Optional. An object literal that contains one or more of the following properties.
-         *        coercionType: If text, the current style is applied in Outlook Web App and Outlook. 
-         *        If the field is an HTML editor, only the text data is inserted, even if the data is HTML. 
-         *        If html and the field supports HTML (the subject doesn't), the current style is applied in Outlook Web App and the default style is 
-         *        applied in Outlook. If the field is a text field, an InvalidDataFormat error is returned. 
-         *        If coercionType is not set, the result depends on the field: if the field is HTML then HTML is used; 
-         *        if the field is text, then plain text is used.
-         */
-        setSelectedDataAsync(data: string, options?: CoercionTypeOptions): void;
+        setSelectedDataAsync(data: string, callback: (asyncResult: Office.AsyncResult<void>) => void): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: This callback should be optional (and the parameter description should match).

File: types/office-js-preview/index.d.ts [hunk 7]
```diff
@@ -18009,31 +16323,7 @@ declare namespace Office {
...
-         * <table>
-         *   <tr><td>{@link https://docs.microsoft.com/outlook/add-ins/understanding-outlook-add-in-permissions | Minimum permission level}</td><td>ReadItem</td></tr>
-         *   <tr><td>{@link https://docs.microsoft.com/outlook/add-ins/#extension-points | Applicable Outlook mode}</td><td>Message Read</td></tr>
-         * </table>
-         *
-         * @param callback - When the method completes, the function passed in the callback parameter is called with a single parameter of 
-         *                 type Office.AsyncResult.
-         */
-        loadCustomPropertiesAsync(callback: (result: Office.AsyncResult<CustomProperties>) => void): void;
+        loadCustomPropertiesAsync(callback: (asyncResult: Office.AsyncResult<CustomProperties>) => void, userContext?: any): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Shouldn't this callback be at the end of the parameter list? And optional?

File: types/office-js-preview/index.d.ts [hunk 8]
```diff
@@ -18162,7 +16434,7 @@ declare namespace Office {
          *
          * @param options - Optional. An object literal that contains one or more of the following properties.
          *        asyncContext: Developers can provide any object they wish to access in the callback method.
-         * @param callback - Optional. When the method completes, the function passed in the callback parameter is called with a single parameter of 
+         * @param callback - When the method completes, the function passed in the callback parameter is called with a single parameter of
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Don't remove the "Optional".

File: types/office-js-preview/index.d.ts [hunk 9]
```diff
@@ -18191,22 +16463,7 @@ declare namespace Office {
...
-         * [Api set: Mailbox 1.1]
-         *
-         * @remarks
-         * <table>
-         *   <tr><td>{@link https://docs.microsoft.com/outlook/add-ins/understanding-outlook-add-in-permissions | Minimum permission level}</td><td>ReadItem</td></tr>
-         *   <tr><td>{@link https://docs.microsoft.com/outlook/add-ins/#extension-points | Applicable Outlook mode}</td><td>Compose</td></tr>
-         * </table>
-         */
-        getAsync(): void;
+        getAsync(callback: (asyncResult: Office.AsyncResult<string>) => void): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Shouldn't this still be optional?

File: types/office-js-preview/index.d.ts [hunk 10]
```diff
@@ -18751,110 +16942,33 @@ declare namespace Office {
...
-         *
-         * *Version differences*
-         *
-         * When you use the makeEwsRequestAsync method in mail apps running in Outlook versions earlier than version 15.0.4535.1004, you should set 
-         * the encoding value to ISO-8859-1.
-         *
-         * `<?xml version="1.0" encoding="iso-8859-1"?>`
+         * @param userContext - Optional. Any state data that is passed to the asynchronous method.
+         */
+        getCallbackTokenAsync(callback: (asyncResult: Office.AsyncResult<string>) => void, userContext?: any): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Shouldn't this callback be at the end of the parameter list? And optional?

File: types/office-js-preview/index.d.ts [hunk 11]
```diff
@@ -18751,110 +16942,33 @@ declare namespace Office {
...
+         * {@link https://docs.microsoft.com/outlook/add-ins/authentication | authenticate the add-in and user with a third-party system}.
+         *
+         * @param callback - When the method completes, the function passed in the callback parameter is called with a single parameter of 
+         *                 type Office.AsyncResult.
+         *                 The token is provided as a string in the `asyncResult.value` property.
+         *                 If there was an error, then the `asyncResult.error` and `asyncResult.diagnostics` properties may provide additional information.
+         * @param userContext - Optional. Any state data that is passed to the asynchronous method.|
          */
-        makeEwsRequestAsync(data: any, callback: (result: Office.AsyncResult<string>) => void, userContext?: any): void;
+        getUserIdentityTokenAsync(callback: (asyncResult: Office.AsyncResult<string>) => void, userContext?: any): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Shouldn't this callback be at the end of the parameter list? And optional?

File: types/office-js-preview/index.d.ts [hunk 12]
```diff
@@ -18906,8 +17020,9 @@ declare namespace Office {
          * @param callback - When the method completes, the function passed in the callback parameter is called with a single parameter of type Office.AsyncResult.
          *                 The `value` property of the result is the XML of the EWS request provided as a string. 
          *                 If the result exceeds 1 MB in size, an error message is returned instead.
+         * @param userContext - Optional. Any state data that is passed to the asynchronous method.
          */
-        makeEwsRequestAsync(data: any, callback: (result: Office.AsyncResult<string>) => void): void;
+        makeEwsRequestAsync(data: any, callback: (asyncResult: Office.AsyncResult<string>) => void, userContext?: any): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Shouldn't this callback be at the end of the parameter list? And optional?

File: types/office-js-preview/index.d.ts [hunk 13]
```diff
@@ -19427,7 +17441,7 @@ declare namespace Office {
          *                 type Office.AsyncResult.
          *                 The `value` property of the result is an array of EmailAddressDetails objects.
          */
-        getAsync(callback: (result: Office.AsyncResult<EmailAddressDetails[]>) => void): void;
+        getAsync(callback: (asyncResult: Office.AsyncResult<EmailAddressDetails[]>) => void): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: This callback should be optional.

File: types/office-js-preview/index.d.ts [hunk 14]
```diff
@@ -19507,13 +17496,12 @@ declare namespace Office {
...
          * @param recipients - The recipients to add to the recipients list.
-         * @param callback - Optional. When the method completes, the function passed in the callback parameter is called with a single parameter of 
+         * @param callback - When the method completes, the function passed in the callback parameter is called with a single parameter of 
          *                 type Office.AsyncResult. 
          *                 If setting the recipients fails the asyncResult.error property will contain a code that indicates any error that occurred 
          *                 while adding the data.
          */
-        setAsync(recipients: (string | EmailUser | EmailAddressDetails)[], callback: (result: Office.AsyncResult<void>) => void): void;
-
+        setAsync(recipients: (string | EmailUser | EmailAddressDetails)[], callback: (asyncResult: Office.AsyncResult<void>) => void): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Should this callback be optional?

File: types/office-js-preview/index.d.ts [hunk 15]
```diff
@@ -20253,7 +18188,7 @@ declare namespace Office {
          * @param callback - When the method completes, the function passed in the callback parameter is called with a single parameter of type Office.AsyncResult.
          *                  The `value` property of the result is the subject of the item.
          */
-        getAsync(callback: (result: Office.AsyncResult<string>) => void): void;
+        getAsync(callback: (asyncResult: Office.AsyncResult<string>) => void): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Should this callback be optional?

File: types/office-js-preview/index.d.ts [hunk 16]
```diff
@@ -20389,7 +18306,7 @@ declare namespace Office {
          * @param callback - When the method completes, the function passed in the callback parameter is called with a single parameter of type Office.AsyncResult.
          *                  The `value` property of the result is a Date object.
          */
-        getAsync(callback: (result: Office.AsyncResult<Date>) => void): void;
+        getAsync(callback: (asyncResult: Office.AsyncResult<Date>) => void): void;
```
Comment (AlexJerabek) [inline types/office-js-preview/index.d.ts]: Should this callback be optional?

**Review by AlexJerabek — CHANGES_REQUESTED** [review]

**Review by AlexJerabek — APPROVED** [review]

**Comment by AlexJerabek** [comment]
Thanks @ElizabethSamuel-MSFT. Looks good to check in now.

**Comment by ElizabethSamuel-MSFT** [comment]
Thanks for the detailed review ;)

**Pull request merged**